### PR TITLE
feat: enable fuel-wallet/sdk to be used directly

### PR DESCRIPTION
**Pull request by eswarasai**
[FRO-83](https://linear.app/fuel-network/issue/FRO-83/enable-fuelweb3-sdk-to-be-use-direct-without-the-need-to-use-the)

### Description

- [x] Added support for `Fuel` initialisation and not having to wait for `window.fuel` to be injected
```
import { Fuel } from '@fuel-wallet/sdk';
const fuel = new Fuel({ name: 'Fuel Wallet' });

// If extension is installed, request is queued and processed as soon as `window.fuel` is available
await fuel.isConnected();
```

### TODO
- [x] `hasWallet` method to detect Fuel extension installed or not

### CODE REVIEW

File: .changeset/sweet-beans-shake.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@fuel-wallet/sdk': minor
+---
+
+Enabled FuelWeb3 SDK to be use directly without the need to wait for the injected script
```
Comment (luizstacio) [inline .changeset/sweet-beans-shake.md:5]:
```suggestion
Enabled FuelWallet SDK to be use directly without the need to wait for the injected script
```

File: packages/sdk/src/FuelWalletConnection.ts [hunk 2]
```diff
@@ -16,16 +16,59 @@ import type { JSONRPCRequest } from 'json-rpc-2.0';
...
+        }
+        if (window.fuel) {
+          executeQueue();
+        }
+      }
+    };
+
+    handleFuelInjected();
+
+    this.timeoutTimer = setInterval(handleFuelInjected, 1000);
```
Comment (luizstacio) [inline packages/sdk/src/FuelWalletConnection.ts:61]: We should abstract this in a better way splitting the functions on utilities or on the body of the class.

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Review by luizstacio — CHANGES_REQUESTED** [review]
- Add some tests using jest or e2e to ensure that the new approach works as expected

File: packages/docs/docs/dev/getting-started.mdx [hunk 3]
```diff
@@ -51,9 +51,16 @@ Alternatively, you can use a [TypeScript reference](https://www.typescriptlang.o
...
 
 ```ts
+import { Fuel } from '@fuel-wallet/sdk';
+
+const fuel = new Fuel();
+
 window.fuel?.connect();
+// OR
+fuel.connect();
 ```
```
Comment (luizstacio) [inline packages/docs/docs/dev/getting-started.mdx:64]: We can break this section in two, from window and from package.

File: packages/sdk/src/Fuel.test.ts [hunk 4]
```diff
@@ -6,12 +6,22 @@ import {
...
   beforeAll(() => {
-    MockConnection.start();
+    // Inject `fuel` on `window` after 500ms
+    setTimeout(() => {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      window.fuel = fuel;
+    }, 500);
+
+    MockConnection.start(fuel);
```
Comment (luizstacio) [inline packages/sdk/src/Fuel.test.ts:24]: We should create another test file for just testing this.

File: packages/sdk/src/FuelWalletConnection.ts [hunk 5]
```diff
@@ -16,16 +16,62 @@ import type { JSONRPCRequest } from 'json-rpc-2.0';
...
+        }
+        clearInterval(this.timeoutTimer);
+      }
+    }
+  }
+
+  constructor() {
+    super();
+    this.handleFuelInjected();
+    this.timeoutTimer = setInterval(this.handleFuelInjected.bind(this), 100);
```
Comment (luizstacio) [inline packages/sdk/src/FuelWalletConnection.ts:64]:
```suggestion
    this._injectionTimeout = setInterval(this.handleFuelInjected.bind(this), 100);
```

File: packages/sdk/src/FuelWalletConnection.ts [hunk 6]
```diff
@@ -16,16 +16,62 @@ import type { JSONRPCRequest } from 'json-rpc-2.0';
...
     const { data: event } = message;
     return (
       message.origin === window.origin && event.target === PAGE_SCRIPT_NAME
     );
   }
 
+  handleFuelInjected() {
+    if (this.retry === 10) {
+      clearInterval(this.timeoutTimer);
+      this.client.rejectAllPendingRequests('fuel not detected on the window!');
```
Comment (luizstacio) [inline packages/sdk/src/FuelWalletConnection.ts:37]:
```suggestion
      this.client.rejectAllPendingRequests('Timeout fuel not detected on the window!');
```

File: packages/sdk/src/FuelWalletConnection.ts [hunk 7]
```diff
@@ -16,16 +16,62 @@ import type { JSONRPCRequest } from 'json-rpc-2.0';
 
 import { WindowConnection } from './connections/WindowConnection';
 import { getTransactionSigner } from './utils/getTransactionSigner';
+import { hasWindow } from './utils/hasWindow';
 
 export class FuelWalletConnection extends WindowConnection {
+  isListenerAdded = false;
+  queue: JSONRPCRequest[] = [];
+  retry = 1;
```
Comment (luizstacio) [inline packages/sdk/src/FuelWalletConnection.ts:24]:
```suggestion
  _retry = 0;
```

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Review by tomiiide — DISMISSED** [review]
LGTM!

File: packages/sdk/src/FuelWalletConnection.ts [hunk 8]
```diff
@@ -16,16 +16,77 @@ import type { JSONRPCRequest } from 'json-rpc-2.0';
...
+    if (!window.fuel) {
+      if (request) {
+        this.queue.push(request);
+      }
+      return;
+    }
+
     if (request) {
       this.postMessage({
         type: MessageTypes.request,
```
Comment (LuizAsFight) [inline packages/sdk/src/FuelWalletConnection.ts:92]:
```suggestion
  async sendRequest(request: JSONRPCRequest | null) {
    if (!request) return;
    
    if (!window.fuel) {
      this.queue.push(request);
    } else {
      this.postMessage({
        type: MessageTypes.request,
```

**Review by LuizAsFight — CHANGES_REQUESTED** [review]

**Review by luizstacio — CHANGES_REQUESTED** [review]

**Review by LuizAsFight — DISMISSED** [review]

File: packages/sdk/src/connections/WindowConnection.ts [hunk 9]
```diff
@@ -9,17 +9,29 @@ import type {
...
+  isListenerAdded = false;
+  queue: JSONRPCRequest[] = [];
+  _retry = 0;
+  _injectionTimeout: NodeJS.Timer;
+  _hasWallet = deferPromise<boolean>();
   connectorName: string;
   private connectors: Array<FuelWalletConnector>;
 
-  constructor(connector: FuelWalletConnector) {
+  constructor(connector: FuelWalletConnector = { name: 'Fuel Wallet' }) {
```
[the remark this answers is not in the document]
Comment (luizstacio) [reply packages/sdk/src/connections/WindowConnection.ts:26]: Yes we should not do this., because is intended to be used by external projects also.

**Review by luizstacio — APPROVED** [review]

**Review by LuizAsFight — APPROVED** [review]

**Pull request merged**